### DISCUSSION TITLE
Skip visibility checks for dynamically generated methods without DynamicMethod

### DISCUSSION THREAD

**Discussion by TwentyFourMinutes** [dotnet/runtime#48024] [Q&A] [answered]
There are similar questions about this topic ob StackOverflow already see: [first](https://stackoverflow.com/questions/36873974/c-sharp-combined-with-msil-jit-skip-verification), [second](https://stackoverflow.com/questions/66094780/skip-visibility-checks-for-dynamically-generated-methods-without-dynamicmethod), [third](https://stackoverflow.com/questions/10012196/why-is-skipping-visibility-checks-only-allowed-for-dynamic-methods?noredirect=1&lq=1) and [fourth](https://stackoverflow.com/questions/13431573/dynamicmethod-is-much-slower-than-compiled-il-function?noredirect=1&lq=1). However I also want to bring up the discussion here.

The essence of the question really is, "How do you ignore accessibility restrictions through the TypeBuilder/MethodBuilder API's". Here is an example of this [problem](https://dotnetfiddle.net/sua9uX).

There are few other questions which raise for me.
- Is there a limitation/restriction which makes this impossible? If so, why is it possible through the `DynamicMethod` API?
- Why doesn't it matter setting the [`skipVerification`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.dynamicmethod.-ctor?view=net-5.0#System_Reflection_Emit_DynamicMethod__ctor_System_String_System_Type_System_Type___System_Type_System_Boolean_) parameter to `true`/`false`. See [this](https://dotnetfiddle.net/BRzgvi) example.
- If this is not possible or will most likely not will not be implemented in the near future, is there any reasonable work around?

**Comment by PathogenDavid** [comment, marked as answer]
You can bypass access checks using [`IgnoresAccessChecksToAttribute`](https://www.strathweb.com/2018/10/no-internalvisibleto-no-problem-bypassing-c-visibility-rules-with-roslyn/). The runtime understands the attribute, but it is not defined in the BCL so you have to define it yourself.

Here's a modified version of your fiddle which includes the attribute: https://dotnetfiddle.net/YPkT7H

------------

As an aside, if that fiddle is the main motivation for you doing this you should look at [`CollectionsMarshal.AsSpan`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan).

**Reply by TwentyFourMinutes** [reply]
I've actually read the exact same article and assumed, that this wouldn't help since I would need to mess with the compiler and the JIT. But this is actually exactly what I was searching for thanks a lot!

But actually no, it was just a simple reproduce able example :).

**Reply by PathogenDavid** [reply]
Glad I could help! Yeah, the attribute has a runtime component and an (unofficial) compiler component since both do access checks. (One upon a time there was a PR to add it to Roslyn proper, but it was unfortunately rejected: https://github.com/dotnet/roslyn/pull/20870)

> But actually no, it was just a simple reproduce able example :).

I figured as much, but thought I'd mention it just in case since that API is relatively new and hidden. (As it probably should be.)

**Reply by TwentyFourMinutes** [reply]
Anyhow do you also know by any chance, similar attributes that would work with .Net FX 4.8? I tried a few attributes such as `ReflectionPermissionAttribute`, `SecurityRulesAttribute`, `SecurityTransparentAttribute` and `AllowPartiallyTrustedCallersAttribute`, sadly non of them seem to work. Here would be a modified example: https://dotnetfiddle.net/3fgZuS

**Reply by PathogenDavid** [reply]
.NET Framework understands `IgnoresAccessChecksToAttribute` too. The issue with it not working on .NET Fiddle is that .NET Fiddle seems to run code without giving it full trust. I tested the fiddle I posted earlier with .NET Framework 4.8 and 4.6.2 locally and it works just fine.

(The concept of [code trust](https://docs.microsoft.com/en-us/dotnet/framework/misc/code-access-security) is an artifact of much older versions of .NET Framework which was deprecated and partially removed in .NET Framework 4 and completely eliminated in .NET Core.)

**Reply by TwentyFourMinutes** [reply]
Oh yeah, bummer. Should have tested it locally as well... However pretty interesting that .Net Fiddle is running the code without full trust 🤔 . Thanks again!

**Reply by PathogenDavid** [reply]
No problem! Yeah, it is a bit strange especially when you consider .NET Fiddle was created after Microsoft deprecated CAS.

**Reply by TahirAhmadov** [reply]
@PathogenDavid Hi, first of all - thanks for posting the solution. I have a minor issue. It looks like if I add the attributes before creating any types, this works. However if I try to add the attributes as I'm creating the types (and methods), it's ignored and I'm getting the exception. It's better to add these attributes when creating methods because at that point I know which assemblies I will need internals access to. 
Is that by design? Is it possible to change?

**Reply by PathogenDavid** [reply]
@TahirAhmadov No problem, glad I could help!

> Is that by design? Is it possible to change?

Unfortunately can't say I would know for either, sorry. That sounds like a bug to me, but it might just be a quirk of how `IgnoresAccessChecksToAttribute` is handled in the context of `System.Reflection.Emit`. I'd recommend [submitting a bug report](https://github.com/dotnet/runtime/issues/new?assignees=&labels=&template=01_bug_report.yml) and see what the area owners say.
